AdGoBye.Plugins/IPlugin.cs

```diff
@@ -18,23 +18,49 @@ public interface IPlugin
+    /// the Plugin has already patched the file.
+    /// </summary>
+    /// <remarks>
+    /// When a Plugin patches a file, the Indexer keeps track of the Plugin that modified that version of the file,
+    /// which non-deterministic and exotically designed Plugins may not benifit from.
+    ///</remarks>
+    /// <returns>Boolean indicating if the Indexer should skip if thought already patched.</returns>
+    bool WantsIndexerTracking();
+
     /// <summary>
     ///     Patch is the main entrypoint to a plugin's operations. Plugins are expected to carry out their respective
     ///     behaviours in this method.
     /// </summary>
-    /// <param name="contentId">A string representing the content's blueprint identifier</param>
-    /// <param name="dataDirectoryPath">The path of the cache directory that contains the __data file</param>
+    /// <param name="fileContainer">Container for the underlying asset being operated on</param>
+    /// <param name="dryRunRequested">Bool representing if the current operation is a dry run, you should only simluate changes if this is true</param>
     /// <returns>A <see cref="EPatchResult" /> that signifies the result of the plugin's patch operation</returns>
-    EPatchResult Patch(string contentId, string dataDirectoryPath);
+    EPatchResult Patch(ref ContentFileContainer fileContainer, bool dryRunRequested);
+
+    /// <summary>
+    ///     Verify is a non-edit stage where Plugins can run environment and validity checks on the asset before
+    ///     operating on it in <see cref="Patch"/>.
+    /// </summary>
+    /// <param name="fileContainer">Container for the underlying asset being operated on</param>
+    /// <returns>A <see cref="EVerifyResult" /> that signifies the result of the plugin's verify operation.
+    ///          Non-<see cref="EVerifyResult.Success"/> returns will skip this Plugin from being executed. </returns>
+    EVerifyResult Verify(ref readonly ContentFileContainer fileContainer);
 
-    /// <param name="contentId">A string representing the content's blueprint identifier</param>
-    /// <param name="dataDirectoryPath">The path of the cache directory that contains the __data file</param>
-    /// <returns>A <see cref="EVerifyResult" /> that signifies the result of the plugin's patch operation</returns>
-    EVerifyResult Verify(string contentId, string dataDirectoryPath);
+    /// <summary>
+    /// Initialize is an optional function ran before <see cref="Verify"/> which Plugins may use to prepare their state
+    /// before patching a world. It's part of the patching loop and therefore may be called multiple times.
+    /// </summary>
+    void Initialize();
+
+    /// <summary>
+    /// PostPatch is an optional function ran after <see cref="Patch"/> in the patch loop which Plugins may use to clean
+    /// their state after patching a world. It's part of the patching loop and therefore may be called multiple times.
+    /// </summary>
+    void PostPatch();
```

Comment:
I brought this one up as a possible area for a notification-styled plugin, looking it over it might be nice to get exactly which patch finished the loop. Otherwise, the plugin doesn't know exactly which patch just completed.

Reply:
We could pass a copy of `Content` for Initialize and PostPatch as that's our exact metadata about the file we'll be operating, we access it as part of checking if the Content is a world.

Reply:
No worries, I've not been working too much on AGB stuff either.

That's more than possible, I just want to avoid passing the content paths to Plugins because I feel like might make them do parsing which we may have done already for them (which is why we give `ContentFileContainer` instead).

Reply:
@DubyaDude Hiya, small bump on this since I figured it went under your radar, this branch is living longer than I'd like.

Reply:
Oh sorry, I assumed you were going forward with passing down the path.
However you decide to do it is fine by me, but it should need passing of some identifier to each of the plugin methods

Reply:
Really ugly, here's what I have staged:
```suggestion
    /// <param name="contentId">ID parsed by the Indexer</param>
    /// <param name="path">Location of the content</param>
    /// <param name="version">Version of the content by </param>
    /// <param name="stableContentName">Hash of the version directory of the content, should be unique</param>
    void PostPatch(string contentId, string path, int version, string stableContentName);
```

Reply:
Hmmm, dont think we'd need the path if were passing in contentid, version, and stable content name. if you'd rather not pass that down there im okay with removing it.

Just be sure this info is given across all method calls (Initialize, Patch, Verify, PostPatch), might be better as its own object class tbh.

Reply:
Yeah, the correctest way to do this would be to pass the `Content` type itself over but it's in the Indexer, so it would need to be extracted into its own package so that both AGB and Plugins can depend on it independently from each other. Then that could just be passed as its own type directly by the Indexer as a sort of context object.

Reply:
Let me experiment with this locally and see how painful this gets.

Reply:
I was being too whiny, this was fine. Does https://github.com/AdGoBye/AdGoBye/blob/b96977774d14eadf4abe938b8b1fa6f216299681/AdGoBye.Plugins/IPlugin.cs look like something you'd imagine?

Reply:
Yea, that looks good to me